VACMS-8362 Homepage redesign - Search functionality

allisonlu opened the pull request:
## Description
Closes [#8362](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/8362) and [#9029](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/9029).

Incorporated the [new Search Input component](https://design.va.gov/components/search-input) from the Design Systems team and added Typeahead logic.

## Screenshots

### Mobile
<img width="378" alt="Screen Shot 2022-05-19 at 5 20 05 PM" src="https://user-images.githubusercontent.com/8542413/169406621-12ccdd3c-1bdf-4b11-893b-c5271bf31b59.png">

### Tablet
<img width="823" alt="Screen Shot 2022-05-19 at 5 20 19 PM" src="https://user-images.githubusercontent.com/8542413/169406664-928b56a1-b949-417b-9ea2-2514a026e5aa.png">

### Desktop
<img width="701" alt="Screen Shot 2022-05-19 at 5 20 50 PM" src="https://user-images.githubusercontent.com/8542413/169406675-042981b8-aedd-409e-af99-36eac9fa3184.png">

### In action
![search-typeahead](https://user-images.githubusercontent.com/8542413/169407706-5f545a54-60c6-43f3-b053-2bf4135a75c1.gif)

## Known issues
- There is a min-width set on the input field, so it overflows on smallest mobile screen (320px)
- There is no z-index value on suggestion list, so it displays "underneath" the image


## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

allisonlu committed:
VACMS-8362:
- search up infrastructure for Search functionality on homepage
- created HomepageSearch component

allisonlu committed:
Merge branch 'feature/VACMS-8340-va.gov-homepage-iteration' into VACMS-8362-homepage-search-functionality

allisonlu committed:
VACMS-8362:
- implement Search input component (from design system) with Typeahead for Homepage Common Tasks section (midpage)

allisonlu committed:
VACMS-8362:
- removed mock suggestions, trying to use Typeahead API instead
- added documentation

allisonlu committed:
VACMS-8362:
- implemented Typeahead logic with Search Input component from design system
- polished functionality when input changes
- removed mock suggestions
- (calling from prod API for demo and testing purposes)

allisonlu committed:
VACMS-8362:
- updated Typeahead API endpoint
- added some documentation

allisonlu committed:
VACMS-8362:
- removed unnecessary styles

ryguyk commented on src/applications/static-pages/homepage/HomepageSearch.jsx:
```diff
@@ -0,0 +1,108 @@
+      return;
+    }
+
+    setTimeout(
```
This `setTimeout` is likely not what we want here. Most likely, it should be replaced with a promise chain (or await).

ryguyk commented on src/applications/static-pages/homepage/HomepageSearch.jsx:
```diff
@@ -0,0 +1,108 @@
+  const [latestSuggestions, setLatestSuggestions] = useState(suggestions);
+
+  // fetch Typeahead suggestions from API
+  const fetchDropDownSuggestions = async inputValue => {
```
This function is doing two things that are likely redundant (and potentially buggy).

1. It returns a value that is used to set the state of the list of suggestions.
2. It sets that state directly, as well.

We don't need both of these to happen. The decision of how to fix it revolves around determining what the purpose of the function is.  Is its purpose to fetch some results and return them? Or is its purpose to fetch some results _and_ update state based on those results? I generally prefer the first (single-responsibility principle). If we go this route, we'll need to keep in mind that it is asynchronous, and we'll have to handle things accordingly when we call this function (see comment below about using a promise chain or await).

ryguyk reviewed (commented):
A few things to note/discuss.

allisonlu committed:
VACMS-8362:
- removed unnecessary component props
- removed state setter from fetch function
- moved return statement to inside of catch block
- replaced setTimeOut with an await Promise

ryguyk reviewed (approved):
Looks great!
